src/bundle/Security/Http/EventListener/CheckTwoFactorCodeReuseListener.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\Http\EventListener;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Psr\Log\LoggerInterface;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeCheckEvent;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeReusedEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use function sha1;
+
+/**
+ * @final
+ */
+class CheckTwoFactorCodeReuseListener implements EventSubscriberInterface
+{
+    public const LISTENER_PRIORITY = 0;
+
+    public function __construct(
+        private readonly EventDispatcherInterface $eventDispatcher,
+        private readonly mixed $cache,
+        private readonly int $cacheDuration,
+        private readonly LoggerInterface|null $logger,
+    ) {
+    }
+
+    public function checkForCodeReuse(TwoFactorCodeCheckEvent $event): void
+    {
+        if (null === $this->cache) {
+            return;
+        }
+
+        $cacheItem = null;
+        $cacheKey = 'scheb_two_factor_code_reuse.'
+            .sha1($event->getUser()->getUserIdentifier().'.'.$event->getCode());
+
+        if ($this->cache instanceof CacheItemPoolInterface) {
+            $cacheItem = $this->cache->getItem($cacheKey);
+            $cacheItem->expiresAfter($this->cacheDuration);
+            $cacheItem->set(true);
+            $this->cache->save($cacheItem);
+        }
+
+        if (!$cacheItem instanceof CacheItemInterface) {
+            if ($this->logger instanceof LoggerInterface) {
+                $this->logger->error('Your logger-cache seems to be configured wrongly! Provide a CacheItemPoolInterface as the cache object if you want to disallow reusing 2FA-codes!');
+            }
+
+            return;
+        }
+
+        if (!$cacheItem->isHit()) {
+            return;
+        }
+
+        $this->eventDispatcher->dispatch(new TwoFactorCodeReusedEvent($event->getUser(), $event->getCode()));
```

Comment:
The early return makes the logic harder to understand. Would actually put the dispatch inside the if, so make it clear what's going on: When there is a cache hit -> dispatch event. 

```suggestion
        if ($cacheItem->isHit()) {
            $this->eventDispatcher->dispatch(new TwoFactorCodeReusedEvent($event->getUser(), $event->getCode()));
        }
```

Reply:
I now know where that came from 😂 

PHP_COdeSniffer complaints about this and phpcbf auto"fixes" it...

```
 56 | ERROR | [x] Use early exit to reduce code nesting.
```

Shall I add a comment to make sure that it stays that way?

Reply:
Well, in that case I'd add a comment to suppress the phpcs error, because I'd value the clarity over "less nesting".